doc/decisions/0_drafts/keyname.md

```diff
@@ -0,0 +1,136 @@
+
+There is a conflict between these two forms in terms of API convenience and efficient execution.
+Generally, the escaped form is more convenient to use, since it is entirely human-readable and just a "normal" string.
+When it comes to execution, however, the unescaped name often makes things more efficient.
+For example, order comparisons between unescaped names are much simpler.
+
+The question now is, which representations should be used by `libelektra-core` and how.
+
+## Constraints
+
+- Because `KeySet` is ordered by name and stores `Key`, the order comparison between the name of two `Key`s must be fast.
+
+## Assumptions
+
+- In most cases the escaped name is used for developer convenience not, because of actual requirements.
+- The most common requirement for using the escaped name is UI: reading names from or displaying them in a user interface (e.g., `kdb` CLI)
+
+## Considered Alternatives
+
+### Only escaped name
+
+Because the escaped name is a simple zero-terminated string, it can be represented as a single `char *`.
+
+Storing the name as a single `char *` would be the most space efficient.
+But resizing would require counting the length every time.
+Therefore, for storage the better solution may be a `char *` and a `size_t`.
+
+However, in the API the name could always be a single `char *`, making for a very easy to use API.
+
+The biggest problem with this approach is that comparing two escaped names is not trivial.
+The comparison needs to account for namespaces, parts and escaping.
+
+Similarly, iterating over the individual parts of a name (and/or manipulating them) is non-trivial, because it requires logic to handle escape sequences.
+
+### Only unescaped name
+
+The unescaped name contains zero bytes.
+It therefore must be represented as a pointer and a size.
+
+This can make for less convenient API, but there are mitigation strategies using additional types.
+Using unescaped names in code can be inconvenient, especially regarding the namespace.
+Without a namespace a name could be written as e.g., `"foo\0bar"`.
+But with a namespace it would be something like `"\1\0\foo\0bar"` and developers would need to remember what namespace `\1` is.
+Using the `KEY_NS_*` constants like this is not easily possible.
+
+Both order and hierarchy comparisons are very simple in this case and can be implemented with little more logic than a few `memcmp`.
+Iterating over the individual parts is also trivial, since all parts are separated by zero bytes.
+
+### Only unescaped name, with separate namespace
+
+To solve the namespace issue of the above solution, the namespace could be separated.
+
+In the API it is an obvious solution.
+Instead of taking a single pointer and size, which receive values like `"\1\0foo\0bar"` and `10`, the API would take a namespace, a pointer, and a size, with values like `KEY_NS_CASCADING`, `"foo\0bar"` and `8`.
+
+Internally, this could be combined into one buffer, stored as a single pointer and size, again.
```

Comment:
Please more details and be more clear how

- the public API would look like and
- how the internal storage would look like.

Also be more clear about the advantages of the separation. Doesn't keyGetNamespace solve the problem?

Reply:
I didn't want to propose specific API here, because that's a lot to think about an discuss. Either I'll create a separate decision for the API, or it'll fall out of another decision (e.g. #4244) as a kind of corollary. The same goes for internal storage, where #4683 and #4676 are important decisions.

What this decision should be about IMO is the question posed in the "Problem" section: "Which representations should be used by `libelektra-core` and how?"

Maybe I could make this more clear, but the difference between the two options in question is:
- "Only unescaped name": Unescaped name is stored internally, and only the unescaped name (as a whole) is used in the API. So `keyGetNamespace (key)` wouldn't be allowed here. It should be replaced with `keyName (key)[0]`
- "Only unescaped name, with separate namespace": Similar, but namespace is separated. It is intentionally left open, whether the unescaped name is stored as one field or two fields (namespace and rest) internally. The important bit is that using the namespace separately in the API is not just allowed, but actually encouraged where it makes sense. A good example is: Something like `keySetName` would probably take the namespace as a separate argument, so the `KEY_NS_*` constants can be used easily.